### PR TITLE
Make underlines stable

### LINKED ISSUE #277: [BUG] Undercurl/dotted/dashed jitters when moving cursor
When moving the cursor over an under-curl as seen in the video, it "jitters", as in the curling moves around.

Applies to:

- undercurl
- dashed underline
- dotted underline

[Screencast from 2024-08-08 23-55-04.webm](https://github.com/user-attachments/assets/ca16dd3e-9608-47f2-a35e-1b005bed138d)

### PR DESCRIPTION
**Pull request by A-Walrus**
Closes #277

### CODE REVIEW

File: src/terminal_box.rs [hunk 1]
```diff
@@ -410,7 +410,7 @@ where
                                 let bottom_offset = style_line_height * 2.0;
 
                                 let full_width = self.end_x - self.start_x;
-                                let mut accu_width = 0.0;
+                                let mut accu_width = -(self.start_x % 4.0);
                                 let mut dot_width = 2.0f32.min(full_width - accu_width);
```
Comment (MoSal) [inline src/terminal_box.rs:414]: With negative `accu_width`, the `.min()` call would not limit the width of the underline to the span.

Same with the other styles below.

Comment (A-Walrus) [reply src/terminal_box.rs]: I reworked them and unified into one function for drawing repeating pattern underlines. I also make sure thatit starts and stops within the span :+1:

File: src/terminal_box.rs [hunk 2]
```diff
@@ -406,72 +406,69 @@ where
                                 renderer.fill_quad(underline2_quad, line_color);
                             }
 
-                            if metadata.flags.contains(Flags::DOTTED_UNDERLINE) {
-                                let bottom_offset = style_line_height * 2.0;
-
+                            let mut draw_repeated = |rects: &[(f32, Option<f32>)]| {
```
Comment (MoSal) [inline src/terminal_box.rs:409]: Maybe add a comment describing the meaning of the tuple fields, or use a struct, although maybe that's an overkill.

File: src/terminal_box.rs [hunk 3]
```diff
@@ -406,72 +406,69 @@ where
...
-                                    dot_width = dot_width.min(full_width - accu_width);
-                                    let pos_offset = mk_pos_offset!(accu_width, bottom_offset);
-                                    let underline_quad =
-                                        mk_quad!(pos_offset, style_line_height, dot_width);
-                                    renderer.fill_quad(underline_quad, line_color);
-                                    accu_width += 2.0 * dot_width;
+                                let pattern_len: f32 = rects.iter().map(|x| x.0).sum();
+                                let mut pos = 0.0;
+                                let mut index = {
+                                    let in_pattern = (self.start_x) % pattern_len;
```
Comment (MoSal) [inline src/terminal_box.rs:414]: Useless parenthesis!

File: src/terminal_box.rs [hunk 4]
```diff
@@ -406,72 +406,69 @@ where
...
-                                while accu_width < full_width {
-                                    dot_width = dot_width.min(full_width - accu_width);
-                                    let pos_offset = mk_pos_offset!(accu_width, bottom_offset);
-                                    let underline_quad =
-                                        mk_quad!(pos_offset, style_line_height, dot_width);
-                                    renderer.fill_quad(underline_quad, line_color);
-                                    accu_width += 2.0 * dot_width;
+                                let pattern_len: f32 = rects.iter().map(|x| x.0).sum();
+                                let mut pos = 0.0;
+                                let mut index = {
```
Comment (MoSal) [inline src/terminal_box.rs:413]: A comment or two in this block might also be helpful.

File: src/terminal_box.rs [hunk 5]
```diff
@@ -406,72 +406,69 @@ where
...
+                                            }
+                                            index = i + 1;
+                                            pos += width;
+                                            break;
+                                        }
+                                    }
+                                    index
+                                };
+                                while pos < full_width {
+                                    let (width, x) = rects[index % rects.len()];
```
Comment (MoSal) [inline src/terminal_box.rs:436]: You forgot to limit the width here. Easy to spot with `./ansi-colors.sh` and dashed underlines ;)

``` rust
let (mut width, x) = rects[index % rects.len()];
width = width.min(full_width - pos);
```

Also `pos` is not very descriptive IMHO. Maybe going back `accu_width` or something similar would be preferred. Feel free to rename `full_width` too if you feel like it.

File: src/terminal_box.rs [hunk 6]
```diff
@@ -406,72 +406,69 @@ where
...
+                                        renderer.fill_quad(underline_quad, line_color);
+                                    }
+                                    pos += width;
+                                    index += 1;
                                 }
+                            };
+
+                            if metadata.flags.contains(Flags::DOTTED_UNDERLINE) {
+                                let bottom_offset = style_line_height * 2.0;
+                                draw_repeated(&[(2.0, Some(bottom_offset)), (3.0, None)]);
```
Comment (MoSal) [inline src/terminal_box.rs:450]: Is that a copy error, or do you prefer `3.0` width gaps over `2.0` ones ;)

File: src/terminal_box.rs [hunk 7]
```diff
@@ -406,72 +406,69 @@ where
...
-
-                                while accu_width < full_width {
-                                    dash_width = dash_width.min(full_width - accu_width);
-                                    let pos_offset = mk_pos_offset!(accu_width, bottom_offset);
-                                    let underline_quad =
-                                        mk_quad!(pos_offset, style_line_height, dash_width);
-                                    renderer.fill_quad(underline_quad, line_color);
-                                    accu_width += dash_width + gap_width;
-                                }
+                                draw_repeated(&[(6.0, Some(bottom_offset)), (3.0, None)]);
```
Comment (MoSal) [inline src/terminal_box.rs:455]: Add a comment describing the tuples passed and/or bring back the descriptive variable names (`dash`, `gap`).

File: src/terminal_box.rs [hunk 8]
```diff
@@ -406,72 +406,69 @@ where
...
 
-                                let full_width = self.end_x - self.start_x;
-                                let mut accu_width = 0.0;
-                                let mut dot_width = 1.0f32.min(full_width - accu_width);
-
-                                while accu_width < full_width {
-                                    dot_width = dot_width.min(full_width - accu_width);
-
-                                    let dot_bottom_offset = match accu_width as u32 % 8 {
+                                let pattern: Vec<_> = (0..8)
```
Comment (MoSal) [inline src/terminal_box.rs:462]: Let's not use the heap for this.

``` rust
// add array to std `use` statements
let pattern: [(f32, Option<f32>); 8] = array::from_fn(|i| match i {
    3..=5 => (1.0, Some(bottom_offset + style_line_height)),
    2 | 6 => (1.0, Some(bottom_offset + 2.0 * style_line_height / 3.0)),
    1 | 7 => (1.0, Some(bottom_offset + 1.0 * style_line_height / 3.0)),
    _ => (1.0, Some(bottom_offset)),
});

```

**Review by MoSal — CHANGES_REQUESTED** [review]
This is almost ready, but not quite.

Left a few comments.

File: src/terminal_box.rs [hunk 9]
```diff
@@ -406,72 +407,79 @@ where
...
-
-                                while accu_width < full_width {
-                                    dash_width = dash_width.min(full_width - accu_width);
-                                    let pos_offset = mk_pos_offset!(accu_width, bottom_offset);
-                                    let underline_quad =
-                                        mk_quad!(pos_offset, style_line_height, dash_width);
-                                    renderer.fill_quad(underline_quad, line_color);
-                                    accu_width += dash_width + gap_width;
-                                }
+                                let dot = (2.0, Some(bottom_offset));
```
Comment (MoSal) [inline src/terminal_box.rs:461]: `dash` and `6.0`.

File: src/terminal_box.rs [hunk 10]
```diff
@@ -406,72 +407,79 @@ where
...
+                                        3..=5 => (1.0, Some(bottom_offset + style_line_height)),
+                                        2 | 6 => (
+                                            1.0,
+                                            Some(bottom_offset + 2.0 * style_line_height / 3.0),
+                                        ),
+                                        1 | 7 => (
+                                            1.0,
+                                            Some(bottom_offset + 1.0 * style_line_height / 3.0),
+                                        ),
+                                        _ => (1.0, Some(bottom_offset)),
```
Comment (MoSal) [inline src/terminal_box.rs:480]: nit.

Should have mentioned this earlier, but I think it's better to be more explicit here now and use:

``` rust
  0 => (1.0, Some(bottom_offset)),
  _ => unreachable!(),
```

**Review by MoSal — CHANGES_REQUESTED** [review]
This should be ready after fixing dashed underlines, and addressing a nitpick.

**Review by MoSal — APPROVED** [review]

**Review by jackpot51 — APPROVED** [review]
Thanks for this work @A-Walrus, and thank you for the reviews @MoSal!

**Pull request merged**